tests/integration/high_availability/test_upgrade.py

```diff
@@ -50,40 +80,112 @@ async def test_pre_upgrade_check(ops_test: OpsTest, mysql_charm_series: str) ->
+        apps=[MYSQL_APP_NAME], status="active", idle_period=30, timeout=TIMEOUT
     )
 
     logger.info("Ensure continuous_writes")
     await ensure_all_units_continuous_writes_incrementing(ops_test)
+
+
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_fail_and_rollback(ops_test, continuous_writes) -> None:
+    logger.info("Get leader unit")
+    leader_unit = await get_leader_unit(ops_test, MYSQL_APP_NAME)
+
+    assert leader_unit is not None, "No leader unit found"
+
+    logger.info("Run pre-upgrade-check action")
+    action = await leader_unit.run_action("pre-upgrade-check")
+    await action.wait()
+
+    fault_charm = f"/tmp/{charm.name}"
+    copy(charm, fault_charm)
+
+    logger.info("Inject dependency fault")
+    await inject_dependency_fault(ops_test, MYSQL_APP_NAME, fault_charm)
+
+    application = ops_test.model.applications[MYSQL_APP_NAME]
+
+    logger.info("Refresh the charm")
+    await application.refresh(path=fault_charm)
+
+    logger.info("Wait for upgrade to fail on leader")
+    await ops_test.model.block_until(
+        lambda: leader_unit.workload_status == "blocked",
+        timeout=TIMEOUT,
+    )
+
+    logger.info("Ensure continuous_writes while in failure state")
+    await ensure_all_units_continuous_writes_incrementing(ops_test)
+
+    logger.info("Re-run pre-upgrade-check action")
+    action = await leader_unit.run_action("pre-upgrade-check")
+    await action.wait()
+
+    logger.info("Re-refresh the charm")
+    await application.refresh(path=charm)
+    logger.info("Wait for upgrade to start")
+    await ops_test.model.block_until(
+        lambda: "waiting" in {unit.workload_status for unit in application.units},
+        timeout=TIMEOUT,
+    )
+    await ops_test.model.wait_for_idle(apps=[MYSQL_APP_NAME], status="active", timeout=TIMEOUT)
+
+    logger.info("Ensure continuous_writes after rollback procedure")
+    await ensure_all_units_continuous_writes_incrementing(ops_test)
+
+    # remove fault charm file
+    os.remove(fault_charm)
+
+
+async def inject_dependency_fault(
+    ops_test: OpsTest, application_name: str, charm_file: Union[str, Path]
+) -> None:
+    """Inject a dependency fault into the mysql charm."""
+    # Open dependency.json and load current charm version
+    with open("src/dependency.json", "r") as dependency_file:
+        current_charm_version = json.load(dependency_file)["charm"]["version"]
+
+    # query running dependency to overwrite with incompatible version
+    relation_data = await get_relation_data(ops_test, application_name, "upgrade")
+
+    loaded_dependency_dict = json.loads(relation_data[0]["application-data"]["dependencies"])
+    loaded_dependency_dict["charm"]["upgrade_supported"] = f">{current_charm_version}"
+    loaded_dependency_dict["charm"]["version"] = f"{int(current_charm_version)+1}"
+
+    # Overwrite dependency.json with incompatible version
+    with zipfile.ZipFile(charm_file, mode="a") as charm_zip:
```

Comment:
why use zipfile?

Reply:
the `.charm` is a zipfile - this is very direct way to overwrite a file within a packed `.charm`